### DISCUSSION TITLE
Can SSH user certificates be renewed?

### DISCUSSION THREAD

**Discussion by anon8675309** [smallstep/certificates#1296] [General] [answered]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to document this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## SSH user cert renewal
I would like documentation on how to renew SSH user certs in a similar way to how SSH host certificates are renewed: non-interactive and without any secrets other than having a signed cert and the corresponding key. This would allow short certificate lifetimes.

Here's what I did find and why it does not meet my requirements:
- a [playbook](https://github.com/smallstep/certificates/blob/master/examples/ansible/smallstep-ssh/tasks/main.yml) that issues a new SSH user cert using a one-time-use token; no renewal information
- [step ssh renew](https://smallstep.com/docs/step-cli/reference/ssh/renew) which seems like it should be the command to use, but it explicitly says "This command cannot be used to renew SSH User Certificates."
- the experimental [nebula provider](https://smallstep.com/docs/step-ca/provisioners#nebula) which is the only provider in [the list](https://smallstep.com/docs/step-ca/provisioners#authorization-scope-by-provisioner) that claims to be able to do a ssh-user-cert-renew; no info on how to renew a user SSH cert, only on how to obtain an X509 cert
- a [comment in a discussion](https://github.com/smallstep/certificates/discussions/579#discussioncomment-755530) which says that, as of 2021, renewal of SSH user certificates was not implemented. Since the [aforementioned table](https://smallstep.com/docs/step-ca/provisioners#authorization-scope-by-provisioner) documents Nebula as being able do this, I believe that the comment is now outdated

I'd expect to find this in the documentation on how to renew SSH user certs in all of the following locations:
- Documentation on the [nebula provider](https://smallstep.com/docs/step-ca/provisioners#nebula) which can reportedly do this
- The [example playbook](https://github.com/smallstep/certificates/blob/master/examples/ansible/smallstep-ssh/tasks/main.yml)
- The `step ssh renew` [documentation](https://smallstep.com/docs/step-cli/reference/ssh/renew) if that command is extended to handle this

Some information about my use case, in case it is helpful:
- Each server has its own SSH user cert, username is the FQDN
- The backup servers trust SSH user certs
- Each backup server has an account for each FQDN that needs to stash their backups there
- I don't want the CA password to live on every server, as that'd let a compromised server obtain a certificate that belongs to another host (among other things)
- SSH user certs are issued at deployment time using ansible.builtin.command, like so:
  - `step ssh certificate {{ansible_fqdn}} .ssh/id_ecdsa --provisioner={{ca_user}}@{{ansible_domain}} --password-file={{ca_password_file}} --no-agent --force`
- Using borg backup, which only supports writing to remote hosts [via SSH](https://borgbackup.readthedocs.io/en/stable/usage/general.html#repository-urls), so mTLS isn't an option short of extending borg backup
- Short certificate lifespans are preferable to long ones
- Having access unexpectedly cut off when a cert expires is bad, and even worse when it's cut off for an automated system

**Comment by dopey** [comment, marked as answer]
Hey @anon8675309 👋 . Thanks for opening the issue!

As you mentioned `step-ca` does not support renewing SSH user certificates (I'm not familiar w/ the Nebula provisioner so maybe @maraino can chime in as to why Nebula would be a special case here).

It has been a while since we made this decision but I'll do my best to remember our reasoning. X509 has a widely used process for revoking certificates - CRL. SSH has the KRL but at the time when we made this decision, KRLs were not widely used. I'd have to do some research to understand how common they are now and how well supported. A certificate that can renew itself using it's own private key is essentially a never expiring credential (especially when there is no easy to use system for revoking and checking revocation status). As a policy we prefer not to issue non-expiring credentials to users or workloads that generally only need that credential for a short period of time.

When describing your use case you mention that your hosts need SSH user certificates (I assume for workflows that involve connecting to other hosts). Our philosophy for this use case is that these SSH User ("bot") certificates should be generated on demand when needed by workflows. One way to do this is by issuing the host a longer lived X509 certificate (using a Cloud Identity provisioner, a JWK provisioner, or a ACME Device Attestation provisioner) and then using this X509 certificate in the X5C provisioner to generate SSH user certificates on demand (this is a pattern we use in our hosted platform). We consider this X509 certificate the hosts "Identity" or "Device" Certificate. This Device Certificate can renew itself (you can set up a daemon to do this), and can generate X509 and SSH credentials for all the workloads that reside on the host.

**Reply by anon8675309** [reply]
Based on your answer and the [X5C docs](https://smallstep.com/docs/step-ca/provisioners/#x5c-x509-certificate), it looks like this is the best solution for me. I appreciate that the documentation warns that this allows any certificate holder to get a certificate for any principal, along with clear instructions on how to restrict this!  The documentation on how to set up the server side is great, so I think I'm all set there.

However when I tried issuing an SSH user certificate, I was told `invalid value 'X5C' for flag '--issuer'`.  Here's the exact command I ran:

```
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=X5C --no-agent --force
```

I also tried no specifying any provisioner so the interactive menu would show up, but the only options I see in that list are JWK, SSHPOP, and ACME... no X5C there. I looked for an example that uses the X5C provisioner in the docs, but couldn't find anything.  I also found [an open ticket](https://github.com/smallstep/docs/issues/75) asking for documentation on X5C.

I fumbled around a little and noticed I can pass in an [--x5c-key](https://smallstep.com/docs/step-cli/reference/ssh/certificate) to `step ssh certificate`.  When I tried this, I was told that `the CA does not have any provisioner configured`, which is simply not true.  To prove this to myself, I ran a command to issue a cert using the `--password-file` and that worked fine.  Below are some of the commands I ran and their outputs (with minor redactions):

```
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=X5C --no-agent --force
invalid value 'X5C' for flag '--issuer'

# I already have an X.509 cert for mTLS for syslog-ng, so I tried using that to get an SSH user cert
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=JWT --x5c-key=/etc/syslog-ng/cert.d/privkey.pem --x5c-cert=/etc/syslog-ng/cert.d/fullchain.pem --no-agent --force
cannot create a new token: the CA does not have any provisioner configured

step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=JWK --x5c-key=/etc/syslog-ng/cert.d/privkey.pem --x5c-cert=/etc/syslog-ng/cert.d/fullchain.pem --no-agent --force
cannot create a new token: the CA does not have any provisioner configured

# Just to make sure I wasn't going mad, I made sure I could find the provisioner when I use a password file
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=admin@`hostname -d` --password-file=ca_password --no-agent --force
✔ Provisioner: admin@domain.tld (JWK) [kid: REDACTED]
✔ CA: https://pki.domain.tld
✔ Private Key: .ssh/id_ecdsa
✔ Public Key: .ssh/id_ecdsa.pub
✔ Certificate: .ssh/id_ecdsa-cert.pub

# I expected if --provisioner=admin@`hostname -d` worked with a password file, it should work with an --x5c-key
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=admin@`hostname -d` --x5c-key=/etc/syslog-ng/cert.d/privkey.pem --x5c-cert=/etc/syslog-ng/cert.d/fullchain.pem --no-agent --force
cannot create a new token: the CA does not have any provisioner configured
```

I find the last two commands perplexing. It's using the same string for the `--provisioner` argument, and it's other arguments that determine whether it decides if it has any provisioners configured or not. Is there any documentation that can help me understand this behavior?


Side note about documentation: the link under the "option" bullet point in the [X5C docs](https://smallstep.com/docs/step-ca/provisioners/#x5c-x509-certificate) is a broken link

**Reply by maraino** [reply]
> However when I tried issuing an SSH user certificate, I was told invalid value 'X5C' for flag '--issuer'....

The issuer is an alias of the provisioner flag, so it's saying that the provisioner named `X5C` does not exist. In fact, I can reproduce that if I use `x5c` instead of `X5C`.

> I also tried no specifying any provisioner so the interactive menu would show up, but the only options I see in that list are JWK, SSHPOP, and ACME... no X5C there...
> When I tried this, I was told that the CA does not have any provisioner configured, which is simply not true...

Do you see the X5C provisioner if you type `step ca provisioner list`?

**Reply by anon8675309** [reply]
I do not see it in the list on the client side or the server side.  This is especially interesting on the server side given that the `step` command to add X5C tells me that it already exists.

```sh
step ca provisioner add x5c-smallstep --type X5C --x5c-root $(step path)/certs/root_ca.crt --ssh-template=$(
step path)/templates/ssh/default.tpl
â CA Configuration: /etc/step-ca/config/ca.json

error storing provisioner: provisioner with name x5c-smallstep already exists


cat $(step path)/templates/ssh/default.tpl
{
    "type": {{ toJson .Type }}, 
    "keyId": {{ toJson .AuthorizationCrt.Subject.CommonName }}, 
    "principals": [
        {{ toJson .AuthorizationCrt.Subject.CommonName }}
    ],  
    "extensions": {{ toJson .Extensions }}, 
    "criticalOptions": {{ toJson .CriticalOptions }}
}
```

**Reply by maraino** [reply]
In your previous command, `step` is trying to modify a ca.json in your `$(step path)`, in this case, in your own local `/etc/step-ca/config/ca.json`. 
* Is that the one you trying to modify? 
* Can you check if that one has the `x5c-smallstep` provisioner? 

The CA always needs a ca.json, but it has two ways to operate, one that holds all the provisioners in the ca.json and another that holds them on the DB. `step` is currently assuming that you have those provisioners in that ca.json. It first tries to check if the online CA is up, then if it has the admin enabled, if not, and your `$(step path)` has a `$(step path)/config/ca.json` it will try to edit that. In this form of operation, you need to restart or reload (HUP signal) the CA to get the new configuration. 
* Does that help?

If the CA has the admin enabled, but it is down (or `step` is not properly configured), and you are running on the same machine as the CA, `step` might assume that it has to edit that `ca.json`, but when the server restarts, it looks for provisioners into the DB, and not the ca.json, it doesn't support both at the same time. We already have an issue open to fix this issue. 
* Do you see `"enableAdmin": true` in that ca.json?

**Reply by anon8675309** [reply]
The `X5C` provisioner is present in `/etc/step-ca/config/ca.json` and I do not see `"enableAdmin": true` anywhere in there. I'm fine with all the provisioners being listed in ca.json, and find that preferable for visibility and ease of use.

I am not sure if I did a SIGHUP/restart previously, but I've done that now and the result is the same: neither my client nor the CA server can find the X5C provisioner. Both machines respond like so:

```sh
step ssh certificate `hostname -f` .ssh/id_ecdsa --provisioner=X5C --x5c-key=/etc/syslog-ng/cert.d/privkey.pem --x5c-cert=/etc/syslog-ng/cert.d/fullchain.pem --no-agent --force
invalid value 'X5C' for flag '--issuer'
```

And here's the relevant section of ca.json:

```sh
grep -B 1 -A 14 X5C /etc/step-ca/config/ca.json
			{
				"type": "X5C",
				"name": "x5c-smallstep",
				"roots": "REDACTED",
				"claims": {
					"enableSSHCA": true,
					"disableRenewal": false,
					"allowRenewalAfterExpiry": false
				},
				"options": {
					"x509": {},
					"ssh": {
						"template": "{\n    \"type\": {{ toJson .Type }}, \n    \"keyId\": {{ toJson .AuthorizationCrt.Subject.CommonName }}, \n    \"principals\": [\n        {{ toJson .AuthorizationCrt.Subject.CommonName }}\n    ],  \n    \"extensions\": {{ toJson .Extensions }}, \n    \"criticalOptions\": {{ toJson .CriticalOptions }}\n}\n"
					}
				}
			}
```

It seems like the command to add the provisioner made all the expected configuration changes, but the provisioner still can not be found.

I tried using strace to get some idea of what the step client is doing. I can see it making a connection to the CA server on port 443, successfully reads and writes some data and exits after a brief exchange.

Meanwhile on the server side, I see the a log entry (using `journalctl -eu step-ca`) that has a matching timestamp and user-agent with a status code 200. This log also showed me the path that was being requested, which was very helpful. I made a request to the `/provisioners` endpoint in my browser and the response does contain the X5C provisioner. The contents of that entry look just like the configuration file snippet I posted above.

So I feel pretty confident that the issue is on the client side. I decided to check to see if I'm running the latest version and found that my `0.23.0` client could be upgraded to `0.23.4`. I pulled the .deb files from GitHub into my apt repo and updated, but the 0.23.4 client behaved exactly the same. I didn't expect such a minor patch to fix this, but I wanted to be sure I wasn't running into something that had already been fixed and released.

* Is there some command I need to run on the client side to enable the X5C provisioner, or should it just see the list of provisioners and try to use it?
* Should I see the step client attempt to load any files related to X5C in my strace output? (because I do not)

**Reply by maraino** [reply]
The only two things I can think of are:
* `step` is configured with a different CA, you can see that running `cat $(step path)/config/defaults.json`
* `step-ca` does not have the SSH keys used to sign ssh certificates. Can you sign a X509 certificate with `step ca certificate ...`? Do you have a block like the following in your ca.json:
```
"ssh": {
  "hostKey": "/etc/step-ca/secrets/ssh_host_ca_key",
  "userKey": "/etc/step-ca/secrets/ssh_user_ca_key",
}
```

If you're willing to share the ca.json, defaults.json and a `tree $(step path)`, anonymizing anything necessary, I might be able to tell you more.

**Reply by anon8675309** [reply]
The correct server is specified in defaults.json and I can obtain x509 certs using the acme protocol. Using step to sign the cert for my SSH host keys also works without any issue (I'm using a token for the initial signing and then it renews itself).

Attached are the config files and trees of both the CA server and a machine with the step client installed.

[redacted_ca.json.txt](https://github.com/smallstep/certificates/files/11007287/redacted_ca.json.txt)
[step-ca.tree.txt](https://github.com/smallstep/certificates/files/11007290/step-ca.tree.txt)

[redacted_defaults.json.txt](https://github.com/smallstep/certificates/files/11007288/redacted_defaults.json.txt)
[step.tree.txt](https://github.com/smallstep/certificates/files/11007289/step.tree.txt)

**Reply by maraino** [reply]
I cannot find anything weird with your configuration, I switched the redacted files with my own ones, and it works. In any case, I've created a new ca and added an X5C provisioner. All the password are "password", there's also a file with it in `secrets/password.txt` Here it is:

[example.tgz](https://github.com/smallstep/certificates/files/11020875/example.tgz)

These are the steps I used:
* Create step path and set it
```console
mkdir /tmp/example
export STEPPATH=/tmp/example
```
* Create CA configuration:
```console
step ca init --ssh --acme
# Follow prompts...
```
* Create X5C PKI:
  * `mkdir x5c && cd x5c`
  * Create root: 
  ```console
  step certificate create --profile root-ca "X5C Root CA" root.crt root.key`
  ```
  * Create intermediate: 
  ```console
  step certificate create --profile intermediate-ca --ca  root.crt --ca-key root.key "X5C Intermediate CA" intermediate.crt intermediate.key
  ```
  * Create leaf for testing, note that it has to have the **bundled intermediate**. If we don't add it, we will need to add the intermediate to the X5C provisioner:
  ```console
  step certificate create --profile leaf --bundle --not-after 8760h --ca intermediate.crt --ca-key intermediate.key --no-password --insecure "X5C Leaf" leaf.crt leaf.key
  ```
* Add X5C provisioner:
```console
cd ..
step ca provisioner add X5C --type X5C --x5c-roots x5c/root.crt --ca-config config/ca.json
```
* Start CA:
```console
step-ca --password-file secrets/password.txt
```
* Test it:
```console
$ step ca certificate www.example.org  example.crt example.key --x5c-cert x5c/leaf.crt --x5c-key x5c/leaf.key
✔ Provisioner: X5C (X5C)
✔ CA: https://localhost
✔ Certificate: example.crt
✔ Private Key: example.key
```

I hope this helps to find out why yours is not working.

**Reply by maraino** [reply]
I forgot that you wanted to test with ssh. I've removed it and add it again with ssh:
```
step ca provisioner remove X5C --ca-config config/ca.json`
step ca provisioner add X5C --ssh --type X5C --x5c-roots x5c/root.crt --ca-config config/ca.json
```
Then I restarted the CA and tested it:
```
step ssh certificate --host --x5c-cert x5c/leaf.crt --x5c-key x5c/leaf.key localhost localhost
```

With the SSH the X5c provisioner looks like this:
```
{
    "type": "X5C",
    "name": "X5C",
    "roots": "LS0...0K",
    "claims": {
        "enableSSHCA": true,
        "disableRenewal": false,
        "allowRenewalAfterExpiry": false
    },
    "options": {
        "x509": {},
        "ssh": {}
    }
}
```

**Reply by anon8675309** [reply]
Thank you. I'll stand up a test instance with a fresh CA using these instructions to make sure I can reproduce your results. Assuming it works as advertised, that should allow me to figure out the difference between my production CA and the newly created on. Once I figure it out, I'll post back here for anyone in the future who is trying to do this same thing.

**Reply by anon8675309** [reply]
If I run the exact same commands as in your example, I am able to replicate your results. However there are a few differences from the example and how I want my system to work.

1. There's yet another root CA just for the X5C provisioner. I'd prefer to use the existing X.509 root CA instead so I can make my infrastructure as simple as possible, but I'll create another root of trust if there is no other option.
2. The last command is an example of signing a host key, not a user key. However it also seems to work without `--host`, so I think that's just a trivial change, but I wanted to make sure it was a typo and not something I'm misunderstanding.
3. The example obtains its X.509 key/cert from X5C instead of ACME, presumably just for simplicity (not having to set up DNS entries and the like), but I am **unable to reproduce the results with a bundled X.509 cert that I obtain from ACME**
4. I want to restrict the principal of the SSH user certs that can be obtained from a client that has an X.509 cert. I think that if I can get the unrestricted version working, adding in the template will be a quick and easy change.

In order to take the example and change it into what I am looking for, here's what I did, using an entirely new CA system for testing.

* Create a directory to hold the configs, keys and certs
```
mkdir -p /etc/step-ca
export STEPPATH=/etc/step-ca
```
* Initialize the CA
```
echo "password" > /etc/step-ca/ca_password
step ca init --ssh --deployment-type=standalone --name=`hostname -d` --dns=`hostname -f` --address=:443 --provisioner=admin@`hostname -d` --password-file=/etc/step-ca/ca_password --with-ca-url=https://`hostname -f` --ssh --acme
```
* Add the X5C provisioner
```
step ca provisioner add X5C --ssh --type X5C --x5c-roots /etc/step-ca/certs/root_ca.crt --ssh-template=/etc/step-ca/templates/ssh/default.tpl --ca-config /etc/step-ca/config/ca.json
```
* Change the max TLS cert lifetime
```
jq '(.authority.provisioners[] | select(.name=="acme")).claims={"maxTLSCertDuration": "8760h"}' $(step path)/config/ca.json | sponge $(step path)/config/ca.json
```
* Set up and start the step-ca service  (following [these instructions](https://smallstep.com/docs/step-ca/certificate-authority-server-production#running-step-ca-as-a-daemon))
* Obtain a X.509 key (from the pki server itself, since this is just a test)
```
step ca certificate `hostname -f` test.pub test.key --issuer=acme -f --not-after=8760h --kty=EC --curve=P-521
```
* Obtain an SSH user key (from the pki server itself, since this is just a test)
```
step ssh certificate `hostname -f` id_ecdsa --x5c-cert test.pub --x5c-key test.key
```

Everything seems to work fine until that last command, which results in a crash. I re-ran it with `STEPDEBUG=1` and am attaching the output here.

[crash.log](https://github.com/smallstep/certificates/files/11114808/crash.log)

Clearly the expected behavior is not a crash, but **should my command using the ACME cert have issued me a SSH user certificate?**

To simplify, I destroyed my test VM and started fresh, but this time I omitted the `--ssh-template` argument when adding the X5C provisioner. The result was the same: acme worked fine, X5C crashed step in the same way.

I also tried including the steps in `Create X5C PKI` to create a root, intermediate and leaf specific to the X5C provisioner and attempting to obtain a SSH user cert failed in the same way.  In this test case, I was able get a SSH host key using the leaf that was set up during the PKI configuration phase.  Dittor for an SSH user key.  Below are the exact commands I used to do so.

```
cd /etc/step-ca/ ; step ssh certificate --host --x5c-cert x5c/leaf.crt --x5c-key x5c/leaf.key --password-file /etc/step-ca/ca_password localhost localhost
rm localhost*
step ssh certificate --x5c-cert x5c/leaf.crt --x5c-key x5c/leaf.key --password-file /etc/step-ca/ca_password --no-agent localhost localhost
```

In all cases, signing an SSH host key worked fine, as did an SSH user key, as long as I used the JWK provisioner. Below are the exact commands I used when verifying this worked as expected.

```
step ssh certificate --host --password-file /etc/step-ca/ca_password --provisioner=admin@`hostname -d` localhost localhost
rm localhost*
step ssh certificate --password-file /etc/step-ca/ca_password --provisioner=admin@`hostname -d` --no-agent localhost2 localhost2
rm localhost*
```

**Reply by anon8675309** [reply]
Are you able to reproduce the crash that I'm seeing?

**Reply by maraino** [reply]
I can reproduce it, let me see what is going on.

**Reply by maraino** [reply]
The problem is that x5c sign doesn't work for P-384 and P-521 curves because there's a bug in a library that we use that returns a list of algorithms that we can use, and we pick the first one, that in this case is ES256, instead of the one specific for the key that should be ES512. 

So I'll fix it and also submit a PR to go-jose to be more specific.

**Reply by maraino** [reply]
@anon8675309 This should fix the problem https://github.com/smallstep/cli/pull/897

**Reply by maraino** [reply]
If you compile `step` from the `master` branch it should work for you, with the version you have it will work for ECDSA P-256, Ed25519, and RSA keys.

**Reply by anon8675309** [reply]
I have compiled the latest code for cli and can confirm that it seems that patch has resolved my issue!

Do you expect this to land in 0.23.5, or is it more likely to be in 0.23.6?

**Reply by maraino** [reply]
It will be in the next release.

**Comment by maraino** [comment]
Hi @anon8675309, the only way to renew an SSH certificate is using SSHPOP provisioner, and this is only enabled for host certificates. So the table that you mention incorrect, @tashian can you take a look to [this](https://smallstep.com/docs/step-ca/provisioners#authorization-scope-by-provisioner).

But you can re-use the same key to get a new certificate using any provisioner, for example like this:
1. Create and sign a new key (it will create mariano, mariano.pub, and mariano-cert.pub):
```
step ssh certificate mariano@smallstep.com mariano
```
2. Re-sign using the same public key:
```
step ssh certificate --sign mariano@smallstep.com mariano.pub
```

**Reply by maraino** [reply]
Of course, if you're using a nebula certificate, you will need the `--nebula-cert` and `--nebula-key` flags too.

**Reply by maraino** [reply]
I've created this https://github.com/smallstep/docs/issues/190
